### ISSUE TITLE
BUG: Unexpected behavior in Overlay with keep_geom_type since Version 1.0.0

### ISSUE THREAD

**Issue by mgoedel** [geopandas/geopandas#3389]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [x] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

#### Code Sample

```python
import geopandas
from shapely import Polygon

invalid_polygon = Polygon([(0, 0), (1, 1), (1, 2), (1, 1), (0, 0)])
square = Polygon(((0,0),(1,0),(1,1),(0,1),(0,0)))
df1 = geopandas.GeoDataFrame({'A':[1,2]}, geometry=[invalid_polygon, square])
df2 = geopandas.GeoDataFrame({'B':[5,6]}, geometry=[square, square])

print(df1.geom_type[0])
print(df2.geom_type[0])
df1_df2 = df1.overlay(df2, keep_geom_type=True)
print(df1_df2.geom_type[0])

```

#### Problem description
Overlay switches the geometry type of the resulting GeoDataFrame entirely. The argument 'keep_geom_type=True' should keep the geometry type of the Geodataframe as Polygon, but in the example above, it changes it completely to Linestring. This Problem occurs since version 1.0.0, when overlay switched from buffer(0) to make_valid (make_valid does change the geometry type of the invalid Polygon, but buffer(0) does not). This results the first geometry of df1 to be of type Linestring, which seems to cause the behavior.

#### Expected Output
df1_df2 should have same geometry type as df1.

**Comment by m-richards** [comment]
Thanks for the report @mgoedel. I guess in the short term we should (partially?) revert #3113, but it would be nice to think about if we could avoid using buffer(0) in this case.

**Comment by theroggy** [comment]
Somewhat related, it would be cleaner if shapely would offer a functionality to do this cleanup like requested here: https://github.com/shapely/shapely/issues/1927

**Comment by martinfleis** [comment]
I suppose it is fine trying to fix this but we should keep in mind that the behaviour of geometry-related operations (like overlay) on invalid geometries is unwarranted and can lead to unexpected outcomes across whole geopandas. What we do in overlay (calling `make_valid` ourselves) is a bit of an exception I am not entirely sure why exists.

**Comment by jorisvandenbossche** [comment]
Some other issues with our implementation that makes this case a bit more difficult to understand:

- We do check if the input has mixed geometry types, but we do that before calling `make_valid` (so in the example here, we don't catch that issue that there are mixed types after make_valid)
- Consequently, our code is not written to be robust against mixed geometry types, and the `keep_geom_type=True` option then just assumes it can take the first geometry type (in this case linestring, and not Polygon, and so therefore keeps the "wrong" geometries, as in this case you would expect to keep the Polygons)

**Comment by theroggy** [comment]
> Some other issues with our implementation that makes this case a bit more difficult to understand:
> 
> * We do check if the input has mixed geometry types, but we do that before calling `make_valid` (so in the example here, we don't catch that issue that there are mixed types after make_valid)
> * Consequently, our code is not written to be robust against mixed geometry types, and the `keep_geom_type=True` option then just assumes it can take the first geometry type (in this case linestring, and not Polygon, and so therefore keeps the "wrong" geometries, as in this case you would expect to keep the Polygons)

These issues should indeed be fixed in #3395

**Comment by StunningShield4504** [comment]
# Try this
import geopandas
from shapely.geometry import Polygon

### Define invalid and valid polygons
invalid_polygon = Polygon([(0, 0), (1, 1), (1, 2), (1, 1), (0, 0)])  # Invalid Polygon
square = Polygon([(0, 0), (1, 0), (1, 1), (0, 1), (0, 0)])  # Valid Polygon

# Create GeoDataFrames
df1 = geopandas.GeoDataFrame({'A': [1, 2]}, geometry=[invalid_polygon, square])
df2 = geopandas.GeoDataFrame({'B': [5, 6]}, geometry=[square, square])

### Print initial geometry types
print(f"df1 initial geom_type[0]: {df1.geom_type[0]}")
print(f"df2 initial geom_type[0]: {df2.geom_type[0]}")

# Apply buffer(0) to fix invalid geometries in df1
df1['geometry'] = df1['geometry'].buffer(0)

### Perform the overlay operation
df1_df2 = df1.overlay(df2, keep_geom_type=True)

### Print resulting geometry type
print(f"Resulting geom_type[0] after overlay: {df1_df2.geom_type[0]}")
